implementations/elixir/ockam/ockam_kafka/lib/interceptor/kafka_interceptor.ex

```diff
@@ -0,0 +1,334 @@
+        _type,
+        state
+      ) do
+    {:error, reason, state}
+  end
+
+  def handle_error({:unsupported_api_version, _api} = reason, _message, _type, state) do
+    {:error, reason, state}
+  end
+
+  def handle_error(reason, message, type, state) do
+    Logger.warn(
+      "Kafka interceptor processing error for type: #{inspect(type)} : #{inspect(reason)}"
+    )
+
+    ## Tunnel messages should still be forwarded even if processing failed
+    {:ok, message, state}
+  end
+
+  @spec process_kafka_response(binary(), state :: map()) ::
+          {:ok, binary(), state :: map()} | {:error, reason :: any(), state :: map()}
+  def process_kafka_response(response, state) do
+    case Parser.parse_response_correlation_id(response) do
+      {:ok, correlation_id, _rest} ->
+        case get_request_header(correlation_id, state) do
+          {:ok, request_header} ->
+            state = cleanup_request_header(correlation_id, state)
+            ## Return error with state to cleanup request header for this correlation id
+            with {:error, reason} <-
+                   process_kafka_response_for_request(request_header, response, state) do
+              {:error, reason, state}
+            end
+
+          {:error, :not_found} ->
+            {:error, {:correlation_id_not_found, correlation_id, response}, state}
+
+          {:error, :not_tracked} ->
+            Logger.info("Correlation id not tracked")
+            {:ok, response, state}
+        end
+
+      {:error, _reason} ->
+        {:error, {:response_header_error, :cannot_parse_correlation_id, response}, state}
+    end
+  end
+
+  @spec process_kafka_response_for_request(RequestHeader.t(), binary(), state :: map()) ::
+          {:ok, binary(), state :: map()} | {:error, reason :: any()}
+  def process_kafka_response_for_request(request_header, response, state) do
+    case Parser.parse_kafka_response_for_request(request_header, response) do
+      {:ok, response_header, response_content_size, response_content} ->
+        case handle_kafka_response(response_header, response_content, state) do
+          {:ok, state} ->
+            {:ok, response, state}
+
+          {:ok, updated_response_content, state} ->
+            ## Simplify reconstruction of response header by using the original header binary
+            ## since we don't change the header we can just reuse it with new response content
+            old_header_binary_size = byte_size(response) - response_content_size
+            <<old_header::binary-size(old_header_binary_size), _response_data::binary>> = response
+
+            case reconstruct_response(old_header, updated_response_content) do
+              {:ok, new_response} ->
+                {:ok, new_response, state}
+
+              {:error, reason} ->
+                {:error, reason}
+            end
+
+          {:error, reason} ->
+            {:error, reason}
+        end
+
+      {:error, reason} ->
+        {:error, reason}
+    end
```

Comment:
No need to change this in this PR, but having `case` nested inside of `case` nested inside of `case` makes the code a bit difficult to follow (similar for `process_kafka_request/2` and `process_kafka_response/2`). There's a refactoring opportunity here for the future